### DISCUSSION TITLE
Deploying to custom server

### DISCUSSION THREAD

**Discussion by daandegooijer** [amannn/next-intl#1556] [Q&A]
Hi,

Im want to bring all our Next.js websites over from vercel to our own hosting solution. This hosting solution does not operate on localhost (or 127.0.0.1) but on another IP address. I can correctly start the Next.js server using the -H parameter.

But everytime i want to visit a locale on the site (i.e. example.com/nl) it gets a loopback to localhost (https://localhost:3000/nl)

The logs show the following:
`✓ Starting...
  ▲ Next.js 14.2.16
  - Local:        http://127.110.72.233:3000
  - Network:      http://127.110.72.233:3000

 ✓ Starting...
 ✓ Ready in 221ms`

`Failed to proxy https://localhost:3000/nl AggregateError
    at internalConnectMultiple (node:net:1114:18)
    at afterConnectMultiple (node:net:1667:5)
    at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
  code: 'ECONNREFUSED',
  [errors]: [
    Error: connect ECONNREFUSED ::1:3000
        at createConnectionError (node:net:1634:14)
        at afterConnectMultiple (node:net:1664:40)
        at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
      errno: -111,
      code: 'ECONNREFUSED',
      syscall: 'connect',
      address: '::1',
      port: 3000
    },`

When i disable next-intl it works.

Does next-intl work without vercel?

**Comment by amannn** [comment]
> But everytime i want to visit a locale on the site (i.e. example.com/nl) it gets a loopback to localhost (https://localhost:3000/nl)

Do you mean you get a redirect where the host is changed to `localhost`? 

I have a feeling this could be an issue with a reverse proxy. Is there an internal proxy that calls the Next.js server with e.g. `localhost` instead of the public URL? Can you share more details about the hosting setup? Also the Node.js version you're using?

Can you try if you get the same error with this middleware instead of the one from `next-intl`?

```tsx
export default function middleware(request: NextRequest) {
  if (request.nextUrl.pathname === '/') {
    const url = new URL('/de', request.url);
    return NextResponse.redirect(url.toString());
  }
}
```

… and then call `/`? Isolating the error a bit further would be really helpful here.
 
> Does next-intl work without vercel?

Absolutely, yes—if you run `npm run start` locally you can verify this.

**Comment by daandegooijer** [comment]
Thanks for the fast reply!

If i go to our homepage https://frontend.foundation.ef2.builders/ It redirects to localhost:3000/nl. Thats because nl is our default language. But instead of going to our root domain, it goes to localhost. 

The server is setup so it has a loopback to another ip address environment=HOST="127.110.72.233",PORT="3000"

Added the above code, the /nl does now work (changed /de to /nl). So it maybe has something to do with the middleware?
https://frontend.foundation.ef2.builders/nl

**Reply by amannn** [reply]
But so on  https://frontend.foundation.ef2.builders/ the provided middleware sample is now running instead of the `next-intl` middleware, right?

It contains the broken redirect that causes the issue:

![Screenshot 2024-11-18 at 18 00 13](https://github.com/user-attachments/assets/873a7377-1dc5-4833-addd-304298380c66)

If you inspect your app locally, you're probably seeing a redirect to `/nl` instead, right?

Can you dig a bit into this and debug why `NextResponse.redirect(…)` causes the broken redirect in this case? Maybe if you add logging for `request.url` in your middleware, you can see the `localhost` header. Maybe the Next.js issue tracker has a bit more on this too …

As far as I know, `next-intl` really just follows [the Next.js docs](https://nextjs.org/docs/app/api-reference/functions/next-response#redirect) here.

**Comment by daandegooijer** [comment]
Thanks for the fast response, gonna dig into it :)

**Comment by daandegooijer** [comment]
Well, delved deeper and deeper. I removed the next-intl all together and added a single redirect in the middleware from the docs and i get the same result.

Asked the hosting provider if they can come up with a solution. Because normal rewrites are also a big part of our projects.

Thanks for the help!

**Reply by amannn** [reply]
That sounds great, would be cool if you could share the resolution then!

I also found these relevant parts in the Next.js source:
1. [Call to `relativizeURL`](https://github.com/vercel/next.js/blob/0714c457376bda1d0821fe06ed27e89e0dd3712a/packages/next/src/server/lib/router-utils/resolve-routes.ts#L631)
2. [Implementation of `relativizeURL`](https://github.com/vercel/next.js/blob/canary/packages/next/src/shared/lib/router/utils/relativize-url.ts)
3. [Implementation of `initUrl`](https://github.com/vercel/next.js/blob/0714c457376bda1d0821fe06ed27e89e0dd3712a/packages/next/src/server/lib/router-utils/resolve-routes.ts#L147)

I think the issue really is that the host is not read correctly by Next.js and therefore not stripped out from the `location` header in the response.

**Comment by daandegooijer** [comment]
Too bad I still dont have a solution. We tried everything but redirects are not working correctly because of the headers. Currently rewriting them in the middleware doesnt work, because i get a 431 with an infinite loop.

So currently my only option is to use Vercel, which i really really don't want to.

**Reply by amannn** [reply]
I think Next.js tries to [read the `host` header](https://github.com/vercel/next.js/blob/f803a950f68aaf0b4cc81a27fec17b2e0bc07cfd/packages/next/src/shared/lib/get-hostname.ts#L16) of the request. If it's not set, it uses `localhost` by default.

Maybe some logging in the middleware could help to see if that's set?

```tsx
function middleware(request: NextRequest) {
  console.log(request.headers.get('host'));
}
```

If not, maybe that's something for your hosting solution to supply.

**Comment by huksley** [comment]
I have the same problem. Either next.js or next-intl incorrectly construct URLs to proxy request. 

Running standalone nextjs behind the reverse proxy with a command 
`HOSTNAME=127.0.0.1 node server.js`
results in errors like `Failed to proxy https://localhost:3000/en`

I have all imaginable headers set
```
proxy_set_header Host \$host;
proxy_set_header X-Real-IP \$remote_addr;
proxy_set_header X-Forwarded-For \$proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Proto \$scheme;
proxy_set_header X-Forwarded-Host \$host;
proxy_set_header X-Forwarded-Port \$server_port;
```
